### PR TITLE
[Operators] add all, any, sum.dim, amax.dim

### PR DESCRIPTION
**Pull request by GwokHiujin**
1. Added all, any operators to support the definitions in the API documentation.
2. Implemented dim for sum and amax, which can support int or int[] inputs.

### CODE REVIEW

File: src/flag_gems/ops/all.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+def all_kernel_dim(
+    inp,
+    out,
+    M,
+    N,
+    BLOCK_M: tl.constexpr,
+    BLOCK_N: tl.constexpr,
+):
+    # Map the program id to the row of inp it should compute.
+    pid = tl.program_id(0) * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]
```
Comment (tongxin) [inline src/flag_gems/ops/all.py]: The naming is kind of misleading... I think that one would normally consider pid to be a single integer for a block program.

File: src/flag_gems/ops/all.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    row_mask = pid < M
+
+    _all = tl.full([BLOCK_M, BLOCK_N], value=1, dtype=tl.int1)
+    for off in range(0, N, BLOCK_N):
+        cols = off + tl.arange(0, BLOCK_N)[None, :]
+        col_mask = cols < N
+        mask = row_mask and col_mask
+
+        a = tl.load(inp + cols, mask, other=1.0)
+        _all = _all and (a != 0)
```
Comment (tongxin) [inline src/flag_gems/ops/all.py:50]: Previous tuning experience tells me looping is likely inferior than direct reduction using tl.reduce.

File: src/flag_gems/ops/all.py (hunk 1 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/all.py]: this is my fault. "pid" is used improperly in many other operators. I'll fix them later.

File: src/flag_gems/ops/all.py (hunk 2 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/all.py:50]: direct reduction requires BLOCK_N to be the next power of 2 of N, which might be beyond the shared memory bound of some gpu platforms.

File: src/flag_gems/ops/all.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    BLOCK_SIZE: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    offset = pid * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)
+    inp_ptrs = inp + offset
+    mask = offset < n_elements
+    inp_val = tl.load(inp_ptrs, mask=mask, other=1.0)
+    all_val = tl.reduce(inp_val != 0, axis=0, combine_fn=reduce_all)
+    mid_ptr = mid + pid
+    mid_mask = pid < mid_size
```
Comment (StrongSpoon) [inline src/flag_gems/ops/all.py]: mid_mask is useless

Comment (StrongSpoon) [reply src/flag_gems/ops/all.py]: so does mid_size

File: src/flag_gems/ops/all.py (hunk 2 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/all.py:50]: Yes, I have repeated this error before: in some cases next_power_of_2(N) will exceed the Triton maximum of 2^20.

File: src/flag_gems/ops/all.py (hunk 3 above)
Comment (GwokHiujin) [reply src/flag_gems/ops/all.py]: done

File: src/flag_gems/ops/vector_norm.py [hunk 4]
```diff
@@ -137,9 +137,12 @@ def vector_norm(x, ord=2, dim=None, keepdim=False, dtype=None):
...
         dim = list(range(x.ndim))
     shape = list(x.shape)
-    dim = [d % x.ndim for d in dim]
+    dim = sorted([d % x.ndim for d in dim])
     order = [i for i in range(x.ndim) if i not in dim] + dim
-    x = x.permute(order).contiguous()
+    if order == shape:
+        x = x.contiguous()
+    else:
+        x = x.permute(order).contiguous()
```
Comment (StrongSpoon) [inline src/flag_gems/ops/vector_norm.py]: undo this

Comment (GwokHiujin) [reply src/flag_gems/ops/vector_norm.py]: done

**Pull request merged**